### ISSUE TITLE
Celery beat should be stateless

### ISSUE THREAD

**Issue by dingobar** [celery/celery#7148]
# Checklist

- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Enhancement%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical enhancement to an existing feature.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22Issue+Type%3A+Enhancement%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed enhancements.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the if the same enhancement was already implemented in the
  master branch.
- [x] I have included all related issues and possible duplicate issues in this issue
      (If there are none, check this box anyway).

# Brief Summary

I would like celery beat to be stateless.

# Design

## Architectural Considerations

None

## Proposed Behavior

I looked in the documentation, as I would like to run Celery beat in a container in k8s. However, I saw this, which makes beat probably unsuited for k8s:
![image](https://user-images.githubusercontent.com/41419288/144875769-9b145b0e-0092-4fdc-8f63-277caa775486.png)

Why is it saving things to the local disk? I think it should use redis or something else for this purpose (at least it should be configurable), otherwise we have to use shared volumes in k8s.

**Comment by auvipy** [comment]
this is a deployment issue right?

**Comment by dingobar** [comment]
Hi, that's right in a way. It's more of a scalability issue though, as having state in the container causes horizontal scaling to be impossible.

**Comment by dingobar** [comment]
In my opinion, celery beat could expose a configurable option to use redis instead of a local flat file.

**Comment by auvipy** [comment]
can you share any example?

**Comment by dingobar** [comment]
Sure, there are two scenarios where stateful celery beat in kubernetes is a bad idea.

1. Node roll-over and scheduled restarts: If a container running beat is restarted, it will forget when the last time it ran its tasks. This is probably not a big issue unless you have tasks that need to run at an exact time, or jobs that cannot run twice, because beat could assume a job was not run or that it missed some jobs.
2. (More critical) Horizontal scaling. Let's say our beat container is under heavy load, and needs to scale up. The way we scale in Kubernetes is by starting another pod running the same process. However, the new pod does not know about the state of the other pod, thus instead of offloading the system, each and every task will instead be run twice.

There is a way to solve this by having a shared read-write drive for the pods to access, but this pattern is not recommended because of potential deadlock issues and a proper data-store should be used instead.

Hope that explained why I think saving the state of celery beat to the local disk is a bad idea, when running in kubernetes.

The short term solution is to run only one instance of beat, which cannot be scaled.

**Comment by Smixi** [comment]
It is stated in the doc :

> You have to ensure only a single scheduler is running for a schedule at a time, otherwise you’d end up with duplicate tasks. Using a centralized approach means the schedule doesn’t have to be synchronized, and the service can operate without using locks

So I guess it's not only about beat being statefull or stateless, you will also need to manually split schedules on instances because either way each beat instance would need to know about others schedulers for firing events only once. 

For instance the [django-celery-beat](https://github.com/celery/django-celery-beat/blob/master/django_celery_beat/schedulers.py) package write entries and schedules inside the database, but as long as you don't have synchronization, you can't run multiple beat instances without task duplication.

Anyway having a shared R/W is a requirement for distributed scheduler

**Comment by dingobar** [comment]
Yeah, in essence it's about offloading the internal flatfile state into a shared datastore. I suggest allowing to use the result backend for this somehow, perhaps starting by supporting only say Redis backend.

**Comment by InfernalAzazel** [comment]
I support, At present, I am developing a project using celery in docker swarm and always lose tasks inexplicably. The code logic is to call the shell terminal to run some scripts, which occupy 1 core of the cpu.

Celery may not be suitable for such long-term resource-consuming tasks
You may need to implement a task scheduling framework yourself

**Comment by frenata** [comment]
FYI, we've solved this *exact* problem by mounting an EFS share and making sure the celerybeat scheduler file is stored there. Of course, that still means we can only run one instance of celerybeat at a time. ;)

**Comment by strugee** [comment]
> EFS share

Clarification: EFS referring to AWS Elastic Filesystem (essentially it's just managed NFS). This way we at least get the ability to manually fail over to another instance if something goes wrong. But it's not automatic.

**Comment by martimors** [comment]
Devs these days typically prefer working with database software or blob-storage services instead of interfacing with file systems directly, hence why I suggested giving the option of offloading the state of beat to that. It would make the entry price much lower for those of us who aren't too familiar with disks and file-systems. Also, as far as I can tell the state file for beat is already a sqlite database. The SQLAlchemy is not a popular backend for Celery, but Redis is, hence why I suggest storing the beat state in Redis could be a nice feature for anyone already using Redis as a broker and result backend for Celery.

**Comment by martimors** [comment]
Should I perhaps move this issue to the new `Discussions` tab?
